modules/rpc/yarpc.go

```diff
@@ -36,66 +35,158 @@ import (
-	config                   yarpcConfig
-	log                      ulog.Log
-	stateMu                  sync.RWMutex
-	inboundMiddlewares       []middleware.UnaryInbound
-	onewayInboundMiddlewares []middleware.OnewayInbound
+	register  registerServiceFunc
+	config    yarpcConfig
+	log       ulog.Log
+	stateMu   sync.RWMutex
+	isRunning bool
 }
 
 var (
-	_dispatcherFn = defaultYarpcDispatcher
+	_dispatcherFn = defaultYARPCDispatcher
 
-	_ service.Module = &YarpcModule{}
+	_ service.Module = &YARPCModule{}
+
+	// Represents the global dispatcher used by all started modules.
+	// The YARPC team advised it to be a 'singleton' to control the lifecycle
+	// of all of the in/out bound traffic.
+	_dispatcher   *yarpc.Dispatcher
+	_dispatcherMu sync.RWMutex
+
+	_configsOnce sync.Once
+	_configs     configCollection
 )
 
-type registerServiceFunc func(module *YarpcModule)
+type registerServiceFunc func(module *YARPCModule)
 
 type yarpcConfig struct {
 	modules.ModuleConfig
-	Bind          string `yaml:"bind"`
-	AdvertiseName string `yaml:"advertiseName"`
+	Bind                    string
+	AdvertiseName           string `yaml:"advertiseName"`
+	inboundMiddleware       []middleware.UnaryInbound
+	onewayInboundMiddleware []middleware.OnewayInbound
+	inbounds                []transport.Inbound
 }
 
-func newYarpcModule(
+// Stores a collection of all modules configs and provides operations that are safe to call from multiple go routines.
+// All the config must share the same AdvertiseName and represent a single service.
+// Configs are referenced and stored in the modules, we store only pointers to them to easily add/remove configs.
+type configCollection struct {
+	sync.RWMutex
+	configs []*yarpcConfig
+}
+
+// Adds the config to collection, the first config sets up the AdvertiseName requirement on consequent additions.
+func (c *configCollection) addConfig(config *yarpcConfig) error {
+	if config == nil {
+		return errors.New("passed a nil config")
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	if len(c.configs) > 0 && config.AdvertiseName != c.configs[0].AdvertiseName {
+		return errs.Errorf("name mismatch, expected: %s; actual: %s",
+			c.configs[0].AdvertiseName,
+			config.AdvertiseName,
+		)
+	}
+
+	c.configs = append(c.configs, config)
+	return nil
+}
+
+// Removes the config from collection.
+// If the config is not found, returns an error, because it means that modules were not started properly.
+func (c *configCollection) removeConfig(config *yarpcConfig) error {
+	c.Lock()
+	defer c.Unlock()
+
+	for i := range c.configs {
+		if c.configs[i] == config {
+			c.configs = append(c.configs[:i], c.configs[i+1:]...)
+			return nil
+		}
+	}
+
+	return errors.New("config not found")
+}
+
+// Merge all the yarpc configs in the collection: transports and middleware are going to be shared.
+// The name comes from the first config in the collection and is the same among all configs.
+func (c *configCollection) mergeConfigs() (conf yarpc.Config, err error) {
```

Comment:
tests? I don't think we have tests with multiple modules using single dispatcher.